File: 3. Behavioral Design Patterns/Template Method/Java/TemplateMethodTemplate.java
// Template Method Pattern – Java Template

abstract class AbstractClass {
    public final void templateMethod() { stepOne(); stepTwo(); hook(); }
    protected abstract void stepOne();
    protected abstract void stepTwo();
    protected void hook() { /* optional override */ }
}

class ConcreteClassA extends AbstractClass {
    protected void stepOne() { System.out.println("A: Step 1"); }
    protected void stepTwo() { System.out.println("A: Step 2"); }
}

class ConcreteClassB extends AbstractClass {
    protected void stepOne() { System.out.println("B: Step 1"); }
    protected void stepTwo() { System.out.println("B: Step 2"); }
    protected void hook() { System.out.println("B: Hook"); }
}

public class TemplateMethodTemplate {
    public static void main(String[] args) { new ConcreteClassA().templateMethod(); new ConcreteClassB().templateMethod(); }
}

File: 3. Behavioral Design Patterns/Template Method/C#/TemplateMethodTemplate.cs
// Template Method Pattern – C# Template

using System;

namespace TemplateMethodTemplate
{
    abstract class AbstractClass
    {
        public void TemplateMethod()
        {
            StepOne();
            StepTwo();
            Hook();
        }
        protected abstract void StepOne();
        protected abstract void StepTwo();
        protected virtual void Hook() { /* optional override */ }
    }

    class ConcreteClassA : AbstractClass
    {
        protected override void StepOne() => Console.WriteLine("A: Step 1");
        protected override void StepTwo() => Console.WriteLine("A: Step 2");
    }

    class ConcreteClassB : AbstractClass
    {
        protected override void StepOne() => Console.WriteLine("B: Step 1");
        protected override void StepTwo() => Console.WriteLine("B: Step 2");
        protected override void Hook() => Console.WriteLine("B: Hook");
    }

    class Program { static void Main() { new ConcreteClassA().TemplateMethod(); new ConcreteClassB().TemplateMethod(); } }
}
